Clippy "explicit_iter_loop" suggestion doesn't compile.

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->

I have the following code:
```rust
const GRID_SIZE: usize = 100;
const BYTES_PER_COLOR: usize = 3;

fn flatten(grid: [[[u8; BYTES_PER_COLOR]; GRID_SIZE]; GRID_SIZE]) -> Vec<u8> {
    let mut flat = Vec::new();

    for row in grid.into_iter() {
        for color in row.iter() {
            flat.extend_from_slice(color);
        }
    }

    flat
}
```

Clippy gives the following lint:
```rust
warning: it is more idiomatic to loop over references to containers instead of using explicit iteration methods
  --> src/pixel_render.rs:33:16
   |
33 |     for row in grid.into_iter() {
   |                ^^^^^^^^^^^^^^^^ help: to write this more concisely, try: `&grid`
   |
   = note: #[warn(explicit_iter_loop)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.166/index.html#explicit_iter_loop
```

After changing the code to the suggestion, it does not compile:
```rust
error[E0277]: the trait bound `&[[[u8; 3]; 100]; 100]: std::iter::Iterator` is not satisfied
  --> src/pixel_render.rs:33:5
   |
33 | /     for row in &grid {
34 | |         for color in row.iter() {
35 | |             flat.extend_from_slice(color);
36 | |         }
37 | |     }
   | |_____^ `&[[[u8; 3]; 100]; 100]` is not an iterator; maybe try calling `.iter()` or a similar method
   |
   = help: the trait `std::iter::Iterator` is not implemented for `&[[[u8; 3]; 100]; 100]`
   = note: required by `std::iter::IntoIterator::into_iter`
```

The real solution here is `grid.iter()`.

My suggestion: somehow check that the suggestion compiles before making it.

--- a/clippy_lints/src/loops.rs
+++ b/clippy_lints/src/loops.rs
@@ -2,7 +2,7 @@ use itertools::Itertools;
 use reexport::*;
 use rustc::hir::*;
 use rustc::hir::def::Def;
-use rustc::hir::def_id; 
+use rustc::hir::def_id;
 use rustc::hir::intravisit::{walk_block, walk_decl, walk_expr, walk_pat, walk_stmt, NestedVisitorMap, Visitor};
 use rustc::hir::map::Node::{NodeBlock, NodeExpr, NodeStmt};
 use rustc::lint::*;
@@ -363,7 +363,7 @@ impl LintPass for Pass {
             EMPTY_LOOP,
             WHILE_LET_ON_ITERATOR,
             FOR_KV_MAP,
-            NEVER_LOOP, 
+            NEVER_LOOP,
             MUT_RANGE_BOUND
         )
     }
@@ -1128,7 +1128,12 @@ fn check_for_loop_arg(cx: &LateContext, pat: &Pat, arg: &Expr, expr: &Expr) {
                 let fn_arg_tys = method_type.fn_sig(cx.tcx).inputs();
                 assert_eq!(fn_arg_tys.skip_binder().len(), 1);
                 if fn_arg_tys.skip_binder()[0].is_region_ptr() {
-                    lint_iter_method(cx, args, arg, method_name);
+                    match cx.tables.expr_ty(&args[0]).sty {
+                        // If the length is greater than 32 no traits are implemented for array and
+                        // therefore we cannot use `&`.
+                        ty::TypeVariants::TyArray(_, size) if const_to_u64(size) > 32 => (),
+                        _ => lint_iter_method(cx, args, arg, method_name)
+                    };
                 } else {
                     let object = snippet(cx, args[0].span, "_");
                     span_lint_and_sugg(
@@ -1319,7 +1324,7 @@ struct MutateDelegate {
 impl<'tcx> Delegate<'tcx> for MutateDelegate {
     fn consume(&mut self, _: NodeId, _: Span, _: cmt<'tcx>, _: ConsumeMode) {
     }
-  
+
     fn matched_pat(&mut self, _: &Pat, _: cmt<'tcx>, _: MatchMode) {
     }
 
@@ -1500,13 +1505,13 @@ impl<'a, 'tcx> Visitor<'tcx> for VarVisitor<'a, 'tcx> {
                     walk_expr(&mut used_visitor, idx);
                     used_visitor.used
                 };
-    
+
                 if index_used {
                     let def = self.cx.tables.qpath_def(seqpath, seqexpr.hir_id);
                     match def {
                         Def::Local(node_id) | Def::Upvar(node_id, ..) => {
                             let hir_id = self.cx.tcx.hir.node_to_hir_id(node_id);
-    
+
                             let parent_id = self.cx.tcx.hir.get_parent(expr.id);
                             let parent_def_id = self.cx.tcx.hir.local_def_id(parent_id);
                             let extent = self.cx.tcx.region_scope_tree(parent_def_id).var_scope(hir_id.local_id);

